src/XR/features/WebXRNearInteraction.ts

```diff
@@ -0,0 +1,590 @@
+import { Scene } from "../../scene";
+import { WebXRControllerComponent } from "../motionController/webXRControllerComponent";
+import { Nullable } from "../../types";
+import { Vector3, Quaternion } from "../../Maths/math.vector";
+import { Ray } from "../../Culling/ray";
+import { PickingInfo } from "../../Collisions/pickingInfo";
+import { WebXRAbstractFeature } from "./WebXRAbstractFeature";
+import { UtilityLayerRenderer } from "../../Rendering/utilityLayerRenderer";
+import { WebXRAbstractMotionController } from "../motionController/webXRAbstractMotionController";
+import { WebXRControllerPointerSelection } from "./WebXRControllerPointerSelection";
+
+/**
+ * Options interface for the near interaction pointer selection module
+ */
+export interface IWebXRNearInteractionOptions {
+    /**
+     * if provided, this scene will be used to render meshes.
+     */
+    customUtilityLayerScene?: Scene;
+    /**
+     * Should meshes created here be added to a utility layer or the main scene
+     */
+    useUtilityLayer?: boolean;
+    /**
+     * the xr input to use with this pointer selection
+     */
+    xrInput: WebXRInput;
+
+    farInteractionFeature: WebXRControllerPointerSelection;
+    /**
+     * Enable pointer selection on all controllers instead of switching between them
+     */
+    enablePointerSelectionOnAllControllers?: boolean;
+    /**
+     * The preferred hand to give the pointer selection to. This will be prioritized when the controller initialize.
+     * If switch is enabled, it will still allow the user to switch between the different controllers
+     */
+    preferredHandedness?: XRHandedness;
+    /**
+     * Disable switching the pointer selection from one controller to the other.
+     * If the preferred hand is set it will be fixed on this hand, and if not it will be fixed on the first controller added to the scene
+     */
+    disableSwitchOnClick?: boolean;
+}
+
+/**
+ * A module that will enable near interaction pointer selection for hands and motion controllers of XR Input Sources
+ */
+export class WebXRNearInteraction extends WebXRAbstractFeature {
+    private static _idCounter = 200;
+
+    private _attachController = (xrController: WebXRInputSource) => {
+        if (this._controllers[xrController.uniqueId]) {
+            // already attached
+            return;
+        }
+        const { hoverIndexMeshTip, pickIndexMeshTip } = this._generateNewHandTipMeshes();
+
+        // get two new meshes
+        this._controllers[xrController.uniqueId] = {
+            xrController,
+            meshUnderPointer: null,
+            pick: null,
+            pickIndexMeshTip,
+            hoverIndexMeshTip,
+            grabRay: new Ray(new Vector3(), new Vector3()),
+            nearInteraction: false,
+            nearGrab: false,
+            nearGrabInProcess: false,
+            id: WebXRNearInteraction._idCounter++,
+        };
+
+        if (this._attachedController) {

```

Comment:
It's a little unclear on what this part does. It looks like it overwrites the attached controller with the current controller ID unless the pointer is only enabled on the other controller?

Reply:
attachedController is populated only when enablePointerSelectionOnAllControllers is not enabled. In case it is enabled, we have more than a single Controller and it is not populated. It is kind of placeholder for knowing which controller is being used if only single controller is enabled in the scene. Also, this is not new logic that I added, it is used from pointerselection code to support more than one controller at a time.

Reply:
Gotcha, thanks for the overview